automation/freebsd_repo/FreeBSD_9.2_Repo/src/usr.bin/look/look.1
.\" Copyright (c) 1990, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)look.1	8.1 (Berkeley) 6/14/93
.\" $FreeBSD: releng/9.2/usr.bin/look/look.1 216370 2010-12-11 08:32:16Z joel $
.\"
.Dd July 17, 2004
.Dt LOOK 1
.Os
.Sh NAME
.Nm look
.Nd display lines beginning with a given string
.Sh SYNOPSIS
.Nm
.Op Fl df
.Op Fl t Ar termchar
.Ar string
.Op Ar
.Sh DESCRIPTION
The
.Nm
utility displays any lines in
.Ar file
which contain
.Ar string
as a prefix.
As
.Nm
performs a binary search, the lines in
.Ar file
must be sorted.
.Pp
If
.Ar file
is not specified, the file
.Pa /usr/share/dict/words
is used, only alphanumeric characters are compared and the case of
alphabetic characters is ignored.
.Pp
The following options are available:
.Bl -tag -width indent
.It Fl d
Dictionary character set and order, i.e., only alphanumeric characters
are compared.
.It Fl f
Ignore the case of alphabetic characters.
.It Fl t
Specify a string termination character, i.e., only the characters
in
.Ar string
up to and including the first occurrence of
.Ar termchar
are compared.
.El
.Sh ENVIRONMENT
The
.Ev LANG , LC_ALL
and
.Ev LC_CTYPE
environment variables affect the execution of the
.Nm
utility.
Their effect is described in
.Xr environ 7 .
.Sh FILES
.Bl -tag -width /usr/share/dict/words -compact
.It Pa /usr/share/dict/words
the dictionary
.El
.Sh EXIT STATUS
The
.Nm
utility exits 0 if one or more lines were found and displayed,
1 if no lines were found, and >1 if an error occurred.
.Sh COMPATIBILITY
The original manual page stated that tabs and blank characters participated
in comparisons when the
.Fl d
option was specified.
This was incorrect and the current man page matches the historic
implementation.
.Sh SEE ALSO
.Xr grep 1 ,
.Xr sort 1
.Sh HISTORY
A
.Nm
utility appeared in
.At v7 .
.Sh BUGS
Lines are not compared according to the current locale's collating
order.
Input files must be sorted with
.Ev LC_COLLATE
set to
.Ql C .


automation/freebsd_repo/FreeBSD_9.2_Repo/src/usr.bin/m4/look.c
/*	$OpenBSD: look.c,v 1.10 2002/04/26 16:15:16 espie Exp $	*/

/*
 * Copyright (c) 1989, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * This code is derived from software contributed to Berkeley by
 * Ozan Yigit at York University.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#if 0
#ifndef lint
static char sccsid[] = "@(#)look.c	8.1 (Berkeley) 6/6/93";
#endif /* not lint */
#endif

#include <sys/cdefs.h>
__FBSDID("$FreeBSD: releng/9.2/usr.bin/m4/look.c 216370 2010-12-11 08:32:16Z joel $");

/*
 * look.c
 * Facility: m4 macro processor
 * by: oz
 */

#include <sys/types.h>
#include <stdio.h>
#include <stdlib.h>
#include <stddef.h>
#include <string.h>
#include "mdef.h"
#include "stdd.h"
#include "extern.h"

static void freent(ndptr);

unsigned int
hash(const char *name)
{
	unsigned int h = 0;
	while (*name)
		h = (h << 5) + h + *name++;
	return (h);
}

/*
 * find name in the hash table
 */
ndptr
lookup(const char *name)
{
	ndptr p;
	unsigned int h;

	h = hash(name);
	for (p = hashtab[h % HASHSIZE]; p != nil; p = p->nxtptr)
		if (h == p->hv && STREQ(name, p->name))
			break;
	return (p);
}

/*
 * hash and create an entry in the hash table.
 * The new entry is added in front of a hash bucket.
 */
ndptr
addent(const char *name)
{
	unsigned int h;
	ndptr p;

	h = hash(name);
	p = (ndptr) xalloc(sizeof(struct ndblock));
	p->nxtptr = hashtab[h % HASHSIZE];
	hashtab[h % HASHSIZE] = p;
	p->name = xstrdup(name);
	p->hv = h;
	return p;
}

static void
freent(ndptr p)
{
	free((char *) p->name);
	if (p->defn != null)
		free((char *) p->defn);
	free((char *) p);
}

/*
 * remove an entry from the hashtable
 */
void
remhash(const char *name, int all)
{
	unsigned int h;
	ndptr xp, tp, mp;

	h = hash(name);
	mp = hashtab[h % HASHSIZE];
	tp = nil;
	while (mp != nil) {
		if (mp->hv == h && STREQ(mp->name, name)) {
			mp = mp->nxtptr;
			if (tp == nil) {
				freent(hashtab[h % HASHSIZE]);
				hashtab[h % HASHSIZE] = mp;
			}
			else {
				xp = tp->nxtptr;
				tp->nxtptr = mp;
				freent(xp);
			}
			if (!all)
				break;
		}
		else {
			tp = mp;
			mp = mp->nxtptr;
		}
	}
}
